utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala

```diff
@@ -22,38 +24,27 @@ class BestEffortJsonSchemaEncoder(validationMode: ValidationMode) {
   private val highPriority: PartialFunction[EncodeInput, EncodeOutput] = Map()
 
   final def encodeOrError(value: Any, schema: Schema): Json = {
-    encode(value, schema).valueOr(errors => throw new RuntimeException(errors.toList.mkString(",")))
+    encodeWithJsonValidation(value, schema).valueOr(errors => throw new RuntimeException(errors.toList.mkString(",")))
   }
 
   private def encodeObject(fields: Map[String, _], parentSchema: ObjectSchema): EncodeOutput = {
-    fields
-      .map(field => (field, parentSchema.getPropertySchemas.get(field._1)))
-      .collect {
-        case ((fieldName, null), schema) if isNullableSchema(schema) => Valid((fieldName, Json.Null))
-        case ((fieldName, value), propertySchema) if (propertySchema != null) && notNullOrRequired(fieldName, value, parentSchema) => encode(value, propertySchema, Some(fieldName)).map(fieldName -> _)
-        case ((fieldName, _), null) if !parentSchema.permitsAdditionalProperties() => error(s"Not expected field with name: ${fieldName} for schema: $parentSchema and policy $validationMode does not allow redundant")
-        case ((fieldName, value), null) => Option(parentSchema.getSchemaOfAdditionalProperties) match {
-          case Some(additionalPropertySchema) => encode(value, additionalPropertySchema).map(fieldName -> _)
-          case None =>  Valid(jsonEncoder.encode(value)).map(fieldName -> _)
-        }
+    fields.keys.toList.union(parentSchema.getPropertySchemas.keySet.asScala.toList).distinct.map{ key =>
+      val schema = Option(parentSchema.getPropertySchemas.get(key))
+      val value = fields.getOrElse(key, null)
+      (key, value, fields.contains(key), schema)
```

Comment:
`value, fields.contains(key),` -> these two can be changed to single Option.

Then we will have three fields in pattern-matching (which is still a lot) and match for `None` or `Some(value)`

Reply:
We can't.. Because some can explicit send `null` here :)

Reply:
`Some(null)` won't be possible ?

Reply:
`Option(null)` is None? :)

Reply:
I cane remove 3th param and replace it by explicit ` fields.contains(fieldName)`

Reply:
You was right, fields.get(key) for null returns Some(null). I though it would return None.. :)

Reply:
cool, thanks! looks a lot better now!